### PR TITLE
Limit snippet characters

### LINKED ISSUE #365: Limit Characters for Post Snippet
## Context 
When creating a post, there is a field for you to create a snippet, and it appears you can make it have limitless characters. This causes issues when the snippet would be used in further cases, as it's to lengthy and would need extra formatting

## Ideal solution (What would you like to see implemented?)

A way to limit the snippet field characters, and also make sure that the snippet is a highlight from the post. Also, a feature that automatically suggests lines to be used as snippet would be cool

### PR DESCRIPTION
**Pull request by freedompraise**
## Description
Added functionality to ensure the snippet field has a limit.
The limit used was 255 other suggestions would be appreciated.
For more detail, it closes #365 

## Changes Made

- [x] Added new feature
- [ ] Fixed bug
- [ ] Refactored code
- [ ] Other (please describe):

## Test Plan
No test plan yet, till the logic is confirmed

## Checklist

- [x] I have read the [contributing guidelines](link_to_contributing_guidelines)
- [x] I have added tests to cover my changes and they all pass in addition to the existing tests.
- [ ] I have added documentation for my changes (if appropriate)

### CODE REVIEW

**Comment by jsolly** [comment]
Looks good to me. I added a small CSS update so the error message looks red when there are too many characters.

Could you add a test for adding a snippet that is too long and it is blocked?

**Review by jsolly — CHANGES_REQUESTED** [review]
Add one test for this

**Comment by freedompraise** [comment]
> Add one test for this

Alright

File: app/tests/test_models.py
Comment (jsolly) [inline app/tests/test_models.py:1]: The feature is 'limit snippet characters' so the test should be for the negative case....try to submit a snippet greater than 400 characters. Should be stopped.

Optionally, you could also write a test with a bunch of links that go over 400 characters (but isn't rejected).

File: app/blog/utils.py [hunk 1]
```diff
@@ -87,3 +87,18 @@ def answer_question(
         model=model,
     )
     return response["choices"][0]["text"].strip()
+
+
+from django.core.exceptions import ValidationError
+import re
+
+
+def snippet_validator(value):
```
Comment (jsolly) [inline app/blog/utils.py:96]: If you compile the ragex it's faster for subsequent calls to snippet_validator

The comment about rasing a validation error is redundant because it's obvious that is happening.

```py
def snippet_validator(value):
    max_length = 400
    link_media_regex = re.compile(r"<a.*?/a>|<img.*?/img>|<video.*?/video>|<audio.*?/audio>", flags=re.IGNORECASE)
    value_without_links_media = link_media_regex.sub("", value)
    if len(value_without_links_media) > max_length:
        raise ValidationError(
            f"The snippet cannot have more than {max_length} characters (excluding links and media)."
        )
 ```

File: app/blog/utils.py
Comment (freedompraise) [inline app/blog/utils.py:1]: I'm having circular import errors in the utils file, as a result of the `Post` and `Category` imports. How should this be resolved?

Comment (freedompraise) [reply app/blog/utils.py:1]: For this, as there's no other solution, I had to create a separate file for the validator. In future refactors, other validators could be moved to this file, I guess

**Comment by jsolly** [comment]
I made the following changes.
1 - Move the snippet validation logic to the Post forms.py file. This way we can add the validator without a new migration.
2 - Created a new test_validators.py testfile
3 - Refactored the snippet validator so `max_length` can be a dependency injection. We want to minimize using globals so the test functions can manipulate this value if they need to.

**Comment by jsolly** [comment]
@freedompraise unfortunately, I think the Snippet validator needs more work. I was able to get it to block me when I was well under 400 characters. This is because the text I copy/pasted in had formatting. See the following GIF


Here is a [link to the video](https://www.screencast.com/t/MOiaR33gK) if that GIF doesn't render.

Also added a failing test for you that contains the HTML shown in the video.

**Review by jsolly — CHANGES_REQUESTED** [review]
Please resolve the failing text and check any other edge cases if you can find them.

**Comment by jsolly** [comment]
If you applied that migration you had, you'll have to roll that back because I removed it.

**Comment by jsolly** [comment]
Nice work @freedompraise. Merging

**Review by jsolly — APPROVED** [review]
Approve!
